src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -0,0 +1,196 @@
+            }
+
+            // TODO Remove language specific code: https://github.com/dotnet/roslyn/issues/30276
+            if (document.Project.Language == LanguageNames.VisualBasic && typeChar == '\n')
+            {
+                return new AsyncCompletionData.CommitResult(isHandled: false, 
+                    AsyncCompletionData.CommitBehavior.RaiseFurtherReturnKeyAndTabKeyCommandHandlers);
+            }
+
+            if (item.InsertText.EndsWith(":") && typeChar == ':')
+            {
+                return new AsyncCompletionData.CommitResult(isHandled: false, 
+                    AsyncCompletionData.CommitBehavior.SuppressFurtherTypeCharCommandHandlers);
+            }
+
+            return CommitResultUnhandled;
+        }
+
+        /// <summary>
+        /// This method needs to support custom procesing of commit characters to be on par with the old completion implementation.
+        /// </summary>
+        private bool IsCommitCharacter(char typeChar, ImmutableArray<CharacterSetModificationRule> rules)
+        {
+            // Tab, Enter and Null (call invoke commit) are always a commit character
+            if (typeChar == '\t' || typeChar == '\n' || typeChar == '\0')
+            {
+                return true;
+            }
+
+            foreach (var rule in rules)
+            {
+                switch (rule.Kind)
+                {
+                    case CharacterSetModificationKind.Add:
+                        if (rule.Characters.Contains(typeChar))
+                        {
+                            return true;
+                        }
+
+                        break;
+
+                    case CharacterSetModificationKind.Remove:
+                        if (rule.Characters.Contains(typeChar))
+                        {
+                            return false;
+                        }
+
+                        break;
+
+                    case CharacterSetModificationKind.Replace:
+                        return rule.Characters.Contains(typeChar);
+                }
+            }
+
+            return s_commitChars.Contains(typeChar);
+        }
+
+        private AsyncCompletionData.CommitBehavior CustomCommit(
+            Document document,
+            IFeaturesCustomCommitCompletionProvider provider,
+            ITextView view,
+            ITextBuffer subjectBuffer,
+            RoslynCompletionItem roslynItem,
+            ITrackingSpan applicableSpan,
+            char? commitCharacter,
+            CancellationToken cancellationToken)
+        {
+            bool includesCommitCharacter;
+            if (!subjectBuffer.CheckEditAccess())
+            {
+                // We are on the wrong thread.
+                return AsyncCompletionData.CommitBehavior.None;
+            }
+
+            if (subjectBuffer.EditInProgress)
+            {
+                return AsyncCompletionData.CommitBehavior.None;
+            }
```

Comment:
can we NFW in both of the above cases?  it indicates a pretty severe issue somewhere that should be fixed.